### PR TITLE
Support Cassandra-4.0 for backend storage

### PR DESCRIPTION
**Pull request by michaelsembwever**

### CODE REVIEW

File: src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java
Comment (adejanovski) [inline src/server/src/main/java/io/cassandrareaper/storage/CassandraStorage.java]: Is this really necessary ? I would expect the name to be `datacenters` and not `"datacenters"`

**Review by adejanovski — APPROVED** [review]
LGTM if the `datacenters` field can indeed be retrieved with double quotes in the getter.

**Comment by michaelsembwever** [comment]
> LGTM if the datacenters field can indeed be retrieved with double quotes in the getter.

Yes, thanks. Has been fixed.

**Comment by michaelsembwever** [comment]
The integration tests are still not passing. I think I'll need some help on them…

**Comment by michaelsembwever** [comment]
> 

Reaper's Cassandra-4.0 integration tests fail on `CassandraStorage.hasLeadOnSegment` and the `UPDATE leader SET reaper_instance_id = ?, reaper_instance_host = ?, last_heartbeat = dateof(now()) WHERE leader_id = ? IF reaper_instance_id = ?` statement timing out.

```
WARN  [2018-05-31 07:03:16,273] io.cassandrareaper.service.SegmentRunner: Failed to connect to a coordinator node for segment 98e4da41-64a0-11e8-9cdd-91ca18da821d
! com.datastax.driver.core.exceptions.OperationTimedOutException: [/127.0.0.1:9042] Timed out waiting for server response
! at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onTimeout(RequestHandler.java:772)
! at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1374)
! at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)
! at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:656)
! at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)
! at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
! ... 1 common frames omitted
! Causing: com.datastax.driver.core.exceptions.OperationTimedOutException: [/127.0.0.1:9042] Timed out waiting for server response
! at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:44)
! at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:26)
! at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
! at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:245)
! at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:68)
! at io.cassandrareaper.storage.CassandraStorage.hasLeadOnSegment(CassandraStorage.java:1144)
! at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:684)
! at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:296)
! at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:134)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
! at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
! at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
WARN  [2018-05-31 07:03:17,655] io.cassandrareaper.service.SegmentRunner: Failed to connect to a coordinator node for segment 98e3efe3-64a0-11e8-9cdd-91ca18da821d
! com.datastax.driver.core.exceptions.OperationTimedOutException: [/127.0.0.2:9042] Timed out waiting for server response
! at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onTimeout(RequestHandler.java:772)
! at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1374)
! at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)
! at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:656)
! at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)
! at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
! ... 1 common frames omitted
! Causing: com.datastax.driver.core.exceptions.OperationTimedOutException: [/127.0.0.2:9042] Timed out waiting for server response
! at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:44)
! at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:26)
! at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
! at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:245)
! at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:68)
! at io.cassandrareaper.storage.CassandraStorage.hasLeadOnSegment(CassandraStorage.java:1144)
! at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:684)
! at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:296)
! at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:134)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
! at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
! at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
```

And then …
```
ERROR [2018-05-31 07:04:36,484] io.cassandrareaper.service.RepairRunner: Executing SegmentRunner failed
! java.lang.AssertionError: Could not release lead on segment 98e4da41-64a0-11e8-9cdd-91ca18da821d
! at io.cassandrareaper.storage.CassandraStorage.releaseLead(CassandraStorage.java:1131)
! at io.cassandrareaper.service.SegmentRunner.releaseLead(SegmentRunner.java:1036)
! at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:143)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
! at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
! at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
ERROR [2018-05-31 07:04:37,871] io.cassandrareaper.service.RepairRunner: Executing SegmentRunner failed
! java.lang.AssertionError: Could not release lead on segment 98e3efe3-64a0-11e8-9cdd-91ca18da821d
! at io.cassandrareaper.storage.CassandraStorage.releaseLead(CassandraStorage.java:1131)
! at io.cassandrareaper.service.SegmentRunner.releaseLead(SegmentRunner.java:1036)
! at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:143)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
! at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
! at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
! at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
! at java.util.concurrent.FutureTask.run(FutureTask.java:266)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
! at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
! at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
! at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
! at java.lang.Thread.run(Thread.java:748)
```

**Comment by michaelsembwever** [comment]
Latest problem is Paxos is failing against trunk Cassandra.
While the exception in Reaper shows the LWT timing out, the server is reporting the following exception:
```
WARN  [MutationStage-1] 2018-06-17 13:42:14,356 AbstractLocalAwareExecutorService.java:167 - Uncaught exception on thread Thread[MutationStage-1,5,main]: null
WARN  [MutationStage-1] 2018-06-17 13:42:14,356 AbstractLocalAwareExecutorService.java:168 - Uncaught exception
java.lang.AssertionError: null
	at org.apache.cassandra.service.paxos.PaxosState.<init>(PaxosState.java:49)
	at org.apache.cassandra.db.SystemKeyspace.loadPaxosState(SystemKeyspace.java:1147)
	at org.apache.cassandra.service.paxos.PaxosState.propose(PaxosState.java:107)
	at org.apache.cassandra.service.paxos.ProposeVerbHandler.doPropose(ProposeVerbHandler.java:34)
	at org.apache.cassandra.service.StorageProxy$7.run(StorageProxy.java:559)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.java:162)
	at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecutorService.java:134)
	at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:110)
	at java.lang.Thread.run(Thread.java:748)
```
(note the stack trace does not get printed normally, `logger.warn("Uncaught exception", t);` was added here as line `AbstractLocalAwareExecutorService.java:168`)

This fault is https://issues.apache.org/jira/browse/CASSANDRA-14356

**Comment by michaelsembwever** [comment]
i'm comfortable to merge this to master now.
i need to clean and rebase the commits to one first, but otherwise the patch is good and ready for me.

The trunk backend builds can fail until CASSANDRA-14356 is merged. (i've tested that it fixes the problem.)

**Comment by michaelsembwever** [comment]
> The trunk backend builds can fail until CASSANDRA-14356 is merged. (i've tested that it fixes the problem.)

CASSANDRA-14356 has been merged. The trunk builds are now also mandatory.

**Review by adejanovski — CHANGES_REQUESTED** [review]
Great investigation to find out LWTs were broken in trunk. 
I'd prefer to have the builds on trunk as optional as it could get unstable for reasons not related to Reaper. 
wdyt ?

**Review by adejanovski — APPROVED** [review]
After discussing this, let's merge and see if the tests are stable enough with trunk as mandatory in the coming days/weeks.

**Pull request merged**
